File: wire-schema/src/test/java/com/squareup/wire/internal/protoparser/MessageElementTest.java
/*
 * Copyright (C) 2014 Square, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.squareup.wire.internal.protoparser;

import com.squareup.wire.internal.protoparser.OptionElement.Kind;
import com.squareup.wire.schema.Location;
import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;
import org.junit.Test;

import static com.squareup.wire.schema.Field.Label.ONE_OF;
import static com.squareup.wire.schema.Field.Label.REQUIRED;
import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.Assert.fail;

public class MessageElementTest {
  Location location = Location.get("file.proto");

  @Test public void locationRequired() {
    try {
      MessageElement.builder(null);
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("location");
    }
  }

  @Test public void nameRequired() {
    try {
      MessageElement.builder(location).qualifiedName("Test").build();
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("name");
    }
  }

  @Test public void nameSetsQualifiedName() {
    MessageElement test = MessageElement.builder(location).name("Test").build();
    assertThat(test.name()).isEqualTo("Test");
    assertThat(test.qualifiedName()).isEqualTo("Test");
  }

  @Test public void nullBuilderValuesThrow() {
    try {
      MessageElement.builder(location).name(null);
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("name");
    }
    try {
      MessageElement.builder(location).qualifiedName(null);
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("qualifiedName");
    }
    try {
      MessageElement.builder(location).documentation(null);
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("documentation");
    }
    try {
      MessageElement.builder(location).addField(null);
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("field");
    }
    try {
      MessageElement.builder(location).addFields(null);
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("fields");
    }
    try {
      MessageElement.builder(location).addFields(Collections.<FieldElement>singleton(null));
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("field");
    }
    try {
      MessageElement.builder(location).addType(null);
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("type");
    }
    try {
      MessageElement.builder(location).addTypes(null);
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("types");
    }
    try {
      MessageElement.builder(location).addTypes(Collections.<TypeElement>singleton(null));
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("type");
    }
    try {
      MessageElement.builder(location).addOneOf(null);
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("oneOf");
    }
    try {
      MessageElement.builder(location).addOneOfs(null);
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("oneOfs");
    }
    try {
      MessageElement.builder(location).addOneOfs(Collections.<OneOfElement>singleton(null));
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("oneOf");
    }
    try {
      MessageElement.builder(location).addExtensions((ExtensionsElement) null);
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("extensions");
    }
    try {
      MessageElement.builder(location).addExtensions((Collection<ExtensionsElement>) null);
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("extensions");
    }
    try {
      MessageElement.builder(location).addExtensions(Collections.<ExtensionsElement>singleton(null));
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("extensions");
    }
    try {
      MessageElement.builder(location).addOption(null);
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("option");
    }
    try {
      MessageElement.builder(location).addOptions(null);
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("options");
    }
    try {
      MessageElement.builder(location).addOptions(Collections.<OptionElement>singleton(null));
      fail();
    } catch (NullPointerException e) {
      assertThat(e).hasMessage("option");
    }
  }

  @Test public void emptyToSchema() {
    TypeElement element = MessageElement.builder(location).name("Message").build();
    String expected = "message Message {}\n";
    assertThat(element.toSchema()).isEqualTo(expected);
  }

  @Test public void simpleToSchema() {
    TypeElement element = MessageElement.builder(location)
        .name("Message")
        .addField(FieldElement.builder(location)
            .label(REQUIRED)
            .type("string")
            .name("name")
            .tag(1)
            .build())
        .build();
    String expected = ""
        + "message Message {\n"
        + "  required string name = 1;\n"
        + "}\n";
    assertThat(element.toSchema()).isEqualTo(expected);
  }

  @Test public void addMultipleFields() {
    FieldElement firstName = FieldElement.builder(location)
        .label(REQUIRED)
        .type("string")
        .name("first_name")
        .tag(1)
        .build();
    FieldElement lastName = FieldElement.builder(location)
        .label(REQUIRED)
        .type("string")
        .name("last_name")
        .tag(2)
        .build();
    MessageElement element = MessageElement.builder(location)
        .name("Message")
        .addFields(Arrays.asList(firstName, lastName))
        .build();
    assertThat(element.fields()).hasSize(2);
  }

  @Test public void simpleWithDocumentationToSchema() {
    TypeElement element = MessageElement.builder(location)
        .name("Message")
        .documentation("Hello")
        .addField(FieldElement.builder(location)
            .label(REQUIRED)
            .type("string")
            .name("name")
            .tag(1)
            .build())
        .build();
    String expected = ""
        + "// Hello\n"
        + "message Message {\n"
        + "  required string name = 1;\n"
        + "}\n";
    assertThat(element.toSchema()).isEqualTo(expected);
  }

  @Test public void simpleWithOptionsToSchema() {
    FieldElement field = FieldElement.builder(location)
        .label(REQUIRED)
        .type("string")
        .name("name")
        .tag(1)
        .build();
    TypeElement element = MessageElement.builder(location)
        .name("Message")
        .addField(field)
        .addOption(OptionElement.create("kit", Kind.STRING, "kat"))
        .build();
    String expected = ""
        + "message Message {\n"
        + "  option kit = \"kat\";\n"
        + "\n"
        + "  required string name = 1;\n"
        + "}\n";
    assertThat(element.toSchema()).isEqualTo(expected);
  }

  @Test public void addMultipleOptions() {
    FieldElement field = FieldElement.builder(location)
        .label(REQUIRED)
        .type("string")
        .name("name")
        .tag(1)
        .build();
    OptionElement kitKat = OptionElement.create("kit", Kind.STRING, "kat");
    OptionElement fooBar = OptionElement.create("foo", Kind.STRING, "bar");
    MessageElement element = MessageElement.builder(location)
        .name("Message")
        .addField(field)
        .addOptions(Arrays.asList(kitKat, fooBar))
        .build();
    assertThat(element.options()).hasSize(2);
  }

  @Test public void simpleWithNestedElementsToSchema() {
    TypeElement element = MessageElement.builder(location)
        .name("Message")
        .addField(FieldElement.builder(location)
            .label(REQUIRED)
            .type("string")
            .name("name")
            .tag(1)
            .build())
        .addType(MessageElement.builder(location)
            .name("Nested")
            .addField(FieldElement.builder(location)
                .label(REQUIRED)
                .type("string")
                .name("name")
                .tag(1)
                .build())
            .build())
        .build();
    String expected = ""
        + "message Message {\n"
        + "  required string name = 1;\n"
        + "\n"
        + "  message Nested {\n"
        + "    required string name = 1;\n"
        + "  }\n"
        + "}\n";
    assertThat(element.toSchema()).isEqualTo(expected);
  }

  @Test public void addMultipleTypes() {
    TypeElement nested1 = MessageElement.builder(location).name("Nested1").build();
    TypeElement nested2 = MessageElement.builder(location).name("Nested2").build();
    TypeElement element = MessageElement.builder(location)
        .name("Message")
        .addField(FieldElement.builder(location)
            .label(REQUIRED)
            .type("string")
            .name("name")
            .tag(1)
            .build())
        .addTypes(Arrays.asList(nested1, nested2))
        .build();
    assertThat(element.nestedElements()).hasSize(2);
  }

  @Test public void simpleWithExtensionsToSchema() {
    TypeElement element = MessageElement.builder(location)
        .name("Message")
        .addField(FieldElement.builder(location)
            .label(REQUIRED)
            .type("string")
            .name("name")
            .tag(1)
            .build())
        .addExtensions(ExtensionsElement.create(location, 500, 501))
        .build();
    String expected = ""
        + "message Message {\n"
        + "  required string name = 1;\n"
        + "\n"
        + "  extensions 500 to 501;\n"
        + "}\n";
    assertThat(element.toSchema()).isEqualTo(expected);
  }

  @Test public void addMultipleExtensions() {
    ExtensionsElement fives = ExtensionsElement.create(location, 500, 501);
    ExtensionsElement sixes = ExtensionsElement.create(location, 600, 601);
    MessageElement element = MessageElement.builder(location)
        .name("Message")
        .addField(FieldElement.builder(location)
            .label(REQUIRED)
            .type("string")
            .name("name")
            .tag(1)
            .build())
        .addExtensions(Arrays.asList(fives, sixes))
        .build();
    assertThat(element.extensions()).hasSize(2);
  }

  @Test public void oneOfToSchema() {
    TypeElement element = MessageElement.builder(location)
        .name("Message")
        .addOneOf(OneOfElement.builder()
            .name("hi")
            .addField(FieldElement.builder(location)
                .label(ONE_OF)
                .type("string")
                .name("name")
                .tag(1)
                .build())
            .build())
        .build();
    String expected = ""
        + "message Message {\n"
        + "  oneof hi {\n"
        + "    string name = 1;\n"
        + "  }\n"
        + "}\n";
    assertThat(element.toSchema()).isEqualTo(expected);
  }

  @Test public void addMultipleOneOfs() {
    OneOfElement hi = OneOfElement.builder()
        .name("hi")
        .addField(FieldElement.builder(location)
            .label(ONE_OF)
            .type("string")
            .name("name")
            .tag(1)
            .build())
        .build();
    OneOfElement hey = OneOfElement.builder()
        .name("hey")
        .addField(FieldElement.builder(location)
            .label(ONE_OF)
            .type("string")
            .name("city")
            .tag(2)
            .build())
        .build();
    MessageElement element = MessageElement.builder(location)
        .name("Message")
        .addOneOfs(Arrays.asList(hi, hey))
        .build();
    assertThat(element.oneOfs()).hasSize(2);
  }

  @Test public void multipleEverythingToSchema() {
    FieldElement field1 = FieldElement.builder(location)
        .label(REQUIRED)
        .type("string")
        .name("name")
        .tag(1)
        .build();
    FieldElement field2 = FieldElement.builder(location)
        .label(REQUIRED)
        .type("bool")
        .name("other_name")
        .tag(2)
        .build();
    FieldElement oneOf1Field = FieldElement.builder(location)
        .label(ONE_OF)
        .type("string")
        .name("namey")
        .tag(3)
        .build();
    OneOfElement oneOf1 = OneOfElement.builder()
        .name("thingy")
        .addField(oneOf1Field)
        .build();
    FieldElement oneOf2Field = FieldElement.builder(location)
        .label(ONE_OF)
        .type("string")
        .name("namer")
        .tag(4)
        .build();
    OneOfElement oneOf2 = OneOfElement.builder()
        .name("thinger")
        .addField(oneOf2Field)
        .build();
    ExtensionsElement extensions1 = ExtensionsElement.create(location, 500, 501);
    ExtensionsElement extensions2 = ExtensionsElement.create(location, 503, 503);
    TypeElement nested = MessageElement.builder(location).name("Nested").addField(field1).build();
    OptionElement option = OptionElement.create("kit", Kind.STRING, "kat");
    TypeElement element = MessageElement.builder(location)
        .name("Message")
        .addField(field1)
        .addField(field2)
        .addOneOf(oneOf1)
        .addOneOf(oneOf2)
        .addType(nested)
        .addExtensions(extensions1)
        .addExtensions(extensions2)
        .addOption(option)
        .build();
    String expected = ""
        + "message Message {\n"
        + "  option kit = \"kat\";\n"
        + "\n"
        + "  required string name = 1;\n"
        + "  required bool other_name = 2;\n"
        + "\n"
        + "  oneof thingy {\n"
        + "    string namey = 3;\n"
        + "  }\n"
        + "  oneof thinger {\n"
        + "    string namer = 4;\n"
        + "  }\n"
        + "\n"
        + "  extensions 500 to 501;\n"
        + "  extensions 503;\n"
        + "\n"
        + "  message Nested {\n"
        + "    required string name = 1;\n"
        + "  }\n"
        + "}\n";
    assertThat(element.toSchema()).isEqualTo(expected);
  }

  @Test public void fieldToSchema() {
    FieldElement field = FieldElement.builder(location)
        .label(REQUIRED)
        .type("string")
        .name("name")
        .tag(1)
        .build();
    String expected = "required string name = 1;\n";
    assertThat(field.toSchema()).isEqualTo(expected);
  }

  @Test public void oneOfFieldToSchema() {
    FieldElement field = FieldElement.builder(location)
        .label(ONE_OF)
        .type("string")
        .name("name")
        .tag(1)
        .build();
    String expected = "string name = 1;\n";
    assertThat(field.toSchema()).isEqualTo(expected);
  }

  @Test public void fieldWithDocumentationToSchema() {
    FieldElement field = FieldElement.builder(location)
        .label(REQUIRED)
        .type("string")
        .name("name")
        .tag(1)
        .documentation("Hello")
        .build();
    String expected = ""
        + "// Hello\n"
        + "required string name = 1;\n";
    assertThat(field.toSchema()).isEqualTo(expected);
  }

  @Test public void fieldWithOptionsToSchema() {
    FieldElement field = FieldElement.builder(location)
        .label(REQUIRED)
        .type("string")
        .name("name")
        .tag(1)
        .addOption(OptionElement.create("kit", Kind.STRING, "kat"))
        .build();
    String expected = "required string name = 1 [\n"
        + "  kit = \"kat\"\n"
        + "];\n";
    assertThat(field.toSchema()).isEqualTo(expected);
  }
}

File: wire-schema/src/commonTest/kotlin/com/squareup/wire/schema/internal/parser/MessageElementTest.kt
/*
 * Copyright (C) 2014 Square, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.squareup.wire.schema.internal.parser

import assertk.assertThat
import assertk.assertions.hasSize
import assertk.assertions.isEqualTo
import com.squareup.wire.schema.Field.Label.OPTIONAL
import com.squareup.wire.schema.Field.Label.REPEATED
import com.squareup.wire.schema.Field.Label.REQUIRED
import com.squareup.wire.schema.Location
import com.squareup.wire.schema.internal.MAX_TAG_VALUE
import com.squareup.wire.schema.internal.parser.OptionElement.Kind
import kotlin.test.Test

class MessageElementTest {
  internal var location = Location.get("file.proto")

  @Test
  fun emptyToSchema() {
    val element = MessageElement(
      location = location,
      name = "Message",
    )
    val expected = "message Message {}\n"
    assertThat(element.toSchema()).isEqualTo(expected)
  }

  @Test
  fun simpleToSchema() {
    val element = MessageElement(
      location = location,
      name = "Message",
      fields = listOf(
        FieldElement(
          location = location,
          label = REQUIRED,
          type = "string",
          name = "name",
          tag = 1,
        ),
      ),
    )
    val expected = """
        |message Message {
        |  required string name = 1;
        |}
        |
    """.trimMargin()
    assertThat(element.toSchema()).isEqualTo(expected)
  }

  @Test
  fun addMultipleFields() {
    val firstName = FieldElement(
      location = location,
      label = REQUIRED,
      type = "string",
      name = "first_name",
      tag = 1,
    )
    val lastName = FieldElement(
      location = location,
      label = REQUIRED,
      type = "string",
      name = "last_name",
      tag = 2,
    )
    val element = MessageElement(
      location = location,
      name = "Message",
      fields = listOf(firstName, lastName),
    )
    assertThat(element.fields).hasSize(2)
  }

  @Test
  fun simpleWithDocumentationToSchema() {
    val element = MessageElement(
      location = location,
      name = "Message",
      documentation = "Hello",
      fields = listOf(
        FieldElement(
          location = location,
          label = REQUIRED,
          type = "string",
          name = "name",
          tag = 1,
        ),
      ),
    )
    val expected = """
        |// Hello
        |message Message {
        |  required string name = 1;
        |}
        |
    """.trimMargin()
    assertThat(element.toSchema()).isEqualTo(expected)
  }

  @Test
  fun simpleWithOptionsToSchema() {
    val field = FieldElement(
      location = location,
      label = REQUIRED,
      type = "string",
      name = "name",
      tag = 1,
    )
    val element = MessageElement(
      location = location,
      name = "Message",
      fields = listOf(field),
      options = listOf(OptionElement.create("kit", Kind.STRING, "kat")),
    )
    val expected =
      """message Message {
        |  option kit = "kat";
        |
        |  required string name = 1;
        |}
        |
      """.trimMargin()
    assertThat(element.toSchema()).isEqualTo(expected)
  }

  @Test
  fun addMultipleOptions() {
    val field = FieldElement(
      location = location,
      label = REQUIRED,
      type = "string",
      name = "name",
      tag = 1,
    )
    val kitKat = OptionElement.create("kit", Kind.STRING, "kat")
    val fooBar = OptionElement.create("foo", Kind.STRING, "bar")
    val element = MessageElement(
      location = location,
      name = "Message",
      fields = listOf(field),
      options = listOf(kitKat, fooBar),
    )
    assertThat(element.options).hasSize(2)
  }

  @Test
  fun simpleWithNestedElementsToSchema() {
    val element = MessageElement(
      location = location,
      name = "Message",
      fields = listOf(
        FieldElement(
          location = location,
          label = REQUIRED,
          type = "string",
          name = "name",
          tag = 1,
        ),
      ),
      nestedTypes = listOf(
        MessageElement(
          location = location,
          name = "Nested",
          fields = listOf(
            FieldElement(
              location = location,
              label = REQUIRED,
              type = "string",
              name = "name",
              tag = 1,
            ),
          ),
        ),
      ),
    )
    val expected = """
        |message Message {
        |  required string name = 1;
        |
        |  message Nested {
        |    required string name = 1;
        |  }
        |}
        |
    """.trimMargin()
    assertThat(element.toSchema()).isEqualTo(expected)
  }

  @Test
  fun addMultipleTypes() {
    val nested1 = MessageElement(
      location = location,
      name = "Nested1",
    )
    val nested2 = MessageElement(
      location = location,
      name = "Nested2",
    )
    val element = MessageElement(
      location = location,
      name = "Message",
      fields = listOf(
        FieldElement(
          location = location,
          label = REQUIRED,
          type = "string",
          name = "name",
          tag = 1,
        ),
      ),
      nestedTypes = listOf(nested1, nested2),
    )
    assertThat(element.nestedTypes).hasSize(2)
  }

  @Test
  fun simpleWithExtensionsToSchema() {
    val element = MessageElement(
      location = location,
      name = "Message",
      fields = listOf(
        FieldElement(
          location = location,
          label = REQUIRED,
          type = "string",
          name = "name",
          tag = 1,
        ),
      ),
      extensions = listOf(ExtensionsElement(location = location, values = listOf(500..501))),
    )
    val expected = """
        |message Message {
        |  required string name = 1;
        |
        |  extensions 500 to 501;
        |}
        |
    """.trimMargin()
    assertThat(element.toSchema()).isEqualTo(expected)
  }

  @Test
  fun addMultipleExtensions() {
    val fives = ExtensionsElement(location = location, values = listOf(500..501))
    val sixes = ExtensionsElement(location = location, values = listOf(600..601))
    val element = MessageElement(
      location = location,
      name = "Message",
      fields = listOf(
        FieldElement(
          location = location,
          label = REQUIRED,
          type = "string",
          name = "name",
          tag = 1,
        ),
      ),
      extensions = listOf(fives, sixes),
    )
    assertThat(element.extensions).hasSize(2)
  }

  @Test
  fun oneOfToSchema() {
    val element = MessageElement(
      location = location,
      name = "Message",
      oneOfs = listOf(
        OneOfElement(
          location = location,
          name = "hi",
          fields = listOf(
            FieldElement(
              location = location,
              type = "string",
              name = "name",
              tag = 1,
            ),
          ),
        ),
      ),
    )
    val expected = """
        |message Message {
        |  oneof hi {
        |    string name = 1;
        |  }
        |}
        |
    """.trimMargin()
    assertThat(element.toSchema()).isEqualTo(expected)
  }

  @Test
  fun oneOfWithGroupToSchema() {
    val element = MessageElement(
      location = location,
      name = "Message",
      oneOfs = listOf(
        OneOfElement(
          location = location,
          name = "hi",
          fields = listOf(
            FieldElement(
              location = location,
              type = "string",
              name = "name",
              tag = 1,
            ),
          ),
          groups = listOf(
            GroupElement(
              location = location.at(5, 5),
              name = "Stuff",
              tag = 3,
              fields = listOf(
                FieldElement(
                  location = location.at(6, 7),
                  label = OPTIONAL,
                  type = "int32",
                  name = "result_per_page",
                  tag = 4,
                ),
                FieldElement(
                  location = location.at(7, 7),
                  label = OPTIONAL,
                  type = "int32",
                  name = "page_count",
                  tag = 5,
                ),
              ),
            ),
          ),
        ),
      ),
    )
    // spotless:off because spotless will remove the indents (trailing spaces) in the oneof block.
    val expected = """
        |message Message {
        |  oneof hi {
        |    string name = 1;
        |  
        |    group Stuff = 3 {
        |      optional int32 result_per_page = 4;
        |      optional int32 page_count = 5;
        |    }
        |  }
        |}
        |
    """.trimMargin()
    // spotless:on
    assertThat(element.toSchema()).isEqualTo(expected)
  }

  @Test
  fun addMultipleOneOfs() {
    val hi = OneOfElement(
      location = location,
      name = "hi",
      fields = listOf(
        FieldElement(
          location = location,
          type = "string",
          name = "name",
          tag = 1,
        ),
      ),
    )
    val hey = OneOfElement(
      location = location,
      name = "hey",
      fields = listOf(
        FieldElement(
          location = location,
          type = "string",
          name = "city",
          tag = 2,
        ),
      ),
    )
    val element = MessageElement(
      location = location,
      name = "Message",
      oneOfs = listOf(hi, hey),
    )
    assertThat(element.oneOfs).hasSize(2)
  }

  @Test
  fun reservedToSchema() {
    val element = MessageElement(
      location = location,
      name = "Message",
      reserveds = listOf(
        ReservedElement(location = location, values = listOf(10, 12..14, "foo")),
        ReservedElement(location = location, values = listOf(10)),
        ReservedElement(location = location, values = listOf(12..MAX_TAG_VALUE)),
        ReservedElement(location = location, values = listOf("foo")),
      ),
    )
    val expected = """
        |message Message {
        |  reserved 10, 12 to 14, "foo";
        |  reserved 10;
        |  reserved 12 to max;
        |  reserved "foo";
        |}
        |
    """.trimMargin()
    assertThat(element.toSchema()).isEqualTo(expected)
  }

  @Test
  fun groupToSchema() {
    val element = MessageElement(
      location = location.at(1, 1),
      name = "SearchResponse",
      groups = listOf(
        GroupElement(
          location = location.at(2, 3),
          label = REPEATED,
          name = "Result",
          tag = 1,
          fields = listOf(
            FieldElement(
              location = location.at(3, 5),
              label = REQUIRED,
              type = "string",
              name = "url",
              tag = 2,
            ),
            FieldElement(
              location = location.at(4, 5),
              label = OPTIONAL,
              type = "string",
              name = "title",
              tag = 3,
            ),
            FieldElement(
              location = location.at(5, 5),
              label = REPEATED,
              type = "string",
              name = "snippets",
              tag = 4,
            ),
          ),
        ),
      ),
    )
    val expected = """
        |message SearchResponse {
        |  repeated group Result = 1 {
        |    required string url = 2;
        |    optional string title = 3;
        |    repeated string snippets = 4;
        |  }
        |}
        |
    """.trimMargin()
    assertThat(element.toSchema()).isEqualTo(expected)
  }

  @Test
  fun multipleEverythingToSchema() {
    val field1 = FieldElement(
      location = location.at(1, 2),
      label = REQUIRED,
      type = "string",
      name = "name",
      tag = 2,
    )
    val oneOf1Field1 = FieldElement(
      location = location.at(1, 1),
      type = "string",
      name = "namey",
      tag = 1,
    )
    val oneOf1Field2 = FieldElement(
      location = location.at(2, 1),
      type = "int32",
      name = "aField",
      tag = 5,
    )

    val oneOf1 = OneOfElement(
      location = location,
      name = "thingy",
      fields = listOf(oneOf1Field1, oneOf1Field2),
    )
    val field2 = FieldElement(
      location = location.at(2, 3),
      label = REQUIRED,
      type = "bool",
      name = "other_name",
      tag = 3,
    )
    val oneOf2Field = FieldElement(
      location = location.at(3, 0),
      type = "string",
      name = "namer",
      tag = 4,
    )
    val oneOf2 = OneOfElement(
      location = location,
      name = "thinger",
      fields = listOf(oneOf2Field),
    )
    val extensions1 = ExtensionsElement(location = location.at(5, 0), values = listOf(500..501))
    val extensions2 = ExtensionsElement(location = location.at(6, 2), values = listOf(503))
    val nested = MessageElement(
      location = location.at(7, 1),
      name = "Nested",
      fields = listOf(field1),
    )
    val option = OptionElement.create("kit", Kind.STRING, "kat")
    val element = MessageElement(
      location = location.at(0, 0),
      name = "Message",
      fields = listOf(field1, field2),
      oneOfs = listOf(oneOf1, oneOf2),
      nestedTypes = listOf(nested),
      extensions = listOf(extensions1, extensions2),
      options = listOf(option),
    )
    val expected = """
        |message Message {
        |  option kit = "kat";
        |
        |  oneof thingy {
        |    string namey = 1;
        |    int32 aField = 5;
        |  }
        |
        |  required string name = 2;
        |
        |  required bool other_name = 3;
        |
        |  oneof thinger {
        |    string namer = 4;
        |  }
        |
        |  extensions 500 to 501;
        |  extensions 503;
        |
        |  message Nested {
        |    required string name = 2;
        |  }
        |}
        |
    """.trimMargin()
    assertThat(element.toSchema()).isEqualTo(expected)
  }

  @Test
  fun fieldToSchema() {
    val field = FieldElement(
      location = location,
      label = REQUIRED,
      type = "string",
      name = "name",
      tag = 1,
    )
    val expected = "required string name = 1;\n"
    assertThat(field.toSchema()).isEqualTo(expected)
  }

  @Test
  fun fieldWithDefaultStringToSchemaInProto2() {
    val field = FieldElement(
      location = location,
      label = REQUIRED,
      type = "string",
      name = "name",
      tag = 1,
      defaultValue = "benoît",
    )
    val expected = "required string name = 1 [default = \"benoît\"];\n"
    assertThat(field.toSchema()).isEqualTo(expected)
  }

  @Test
  fun fieldWithDefaultNumberToSchema() {
    val field = FieldElement(
      location = location,
      label = REQUIRED,
      type = "int32",
      name = "age",
      tag = 1,
      defaultValue = "34",
    )
    val expected = "required int32 age = 1 [default = 34];\n"
    assertThat(field.toSchema()).isEqualTo(expected)
  }

  @Test
  fun fieldWithDefaultBoolToSchema() {
    val field = FieldElement(
      location = location,
      label = REQUIRED,
      type = "bool",
      name = "human",
      tag = 1,
      defaultValue = "true",
    )
    val expected = "required bool human = 1 [default = true];\n"
    assertThat(field.toSchema()).isEqualTo(expected)
  }

  @Test
  fun oneOfFieldToSchema() {
    val field = FieldElement(
      location = location,
      type = "string",
      name = "name",
      tag = 1,
    )
    val expected = "string name = 1;\n"
    assertThat(field.toSchema()).isEqualTo(expected)
  }

  @Test
  fun fieldWithDocumentationToSchema() {
    val field = FieldElement(
      location = location,
      label = REQUIRED,
      type = "string",
      name = "name",
      tag = 1,
      documentation = "Hello",
    )
    val expected =
      """// Hello
        |required string name = 1;
        |
      """.trimMargin()
    assertThat(field.toSchema()).isEqualTo(expected)
  }

  @Test
  fun fieldWithOneOptionToSchema() {
    val field = FieldElement(
      location = location,
      label = REQUIRED,
      type = "string",
      name = "name",
      tag = 1,
      options = listOf(OptionElement.create("kit", Kind.STRING, "kat")),
    )
    val expected =
      """required string name = 1 [kit = "kat"];
        |
      """.trimMargin()
    assertThat(field.toSchema()).isEqualTo(expected)
  }

  @Test
  fun fieldWithMoreThanOneOptionToSchema() {
    val field = FieldElement(
      location = location,
      label = REQUIRED,
      type = "string",
      name = "name",
      tag = 1,
      options = listOf(
        OptionElement.create("kit", Kind.STRING, "kat"),
        OptionElement.create("dup", Kind.STRING, "lo"),
      ),
    )
    val expected =
      """required string name = 1 [
        |  kit = "kat",
        |  dup = "lo"
        |];
        |
      """.trimMargin()
    assertThat(field.toSchema()).isEqualTo(expected)
  }

  @Test fun oneOfWithOptions() {
    val expected = """
        |oneof page_info {
        |  option (my_option) = true;
        |
        |  int32 page_number = 2;
        |  int32 result_per_page = 3;
        |}
        |
    """.trimMargin()
    val oneOf = OneOfElement(
      location = location,
      name = "page_info",
      fields = listOf(
        FieldElement(
          location = location.at(4, 5),
          type = "int32",
          name = "page_number",
          tag = 2,
        ),
        FieldElement(
          location = location.at(5, 5),
          type = "int32",
          name = "result_per_page",
          tag = 3,
        ),
      ),
      options = listOf(
        OptionElement.create("my_option", Kind.BOOLEAN, "true", true),
      ),
    )
    assertThat(oneOf.toSchema()).isEqualTo(expected)
  }
}
